tvm/serialization/boc.mdx

```diff
@@ -2,4 +2,142 @@
+- The total number of cells `n` present in the serialization.
+- The number of root cells `k ≤ n` present in the serialization. The root
+cells themselves are $c_1, \ldots, c_{k}$. All other cells present in the bag of
+cells are expected to be reachable by chains of references starting from
+the root cells.
+- The number of absent cells `l ≤ n − k`, which represent cells that are
+actually absent from this bag of cells, but are referred to from it. The
+absent cells themselves are represented by $c_{n−l+1}, \ldots, c_{n}$, and only
+these cells may (and also must) have `r = 7`. Complete bags of cells
+have `l = 0`.
+- The total length in bytes `Ln` of the serialization of all cells. If the index
+is present, `Ln` might not be stored explicitly since it can be recovered as the last entry of the index.
+
+## A TL-B scheme
+
+Only one [serialization scheme of BoCs](https://github.com/ton-blockchain/ton/blob/24dc184a2ea67f9c47042b4104bbb4d82289fac1/crypto/tl/boc.tlb#L25) 
+is used in TON Blockchain:
+```tlb
+serialized_boc#b5ee9c72 has_idx:(## 1) has_crc32c:(## 1) 
+  has_cache_bits:(## 1) flags:(## 2) { flags = 0 }
+  size:(## 3) { size <= 4 }
+  off_bytes:(## 8) { off_bytes <= 8 } 
+  cells:(##(size * 8)) 
+  roots:(##(size * 8)) { roots >= 1 }
+  absent:(##(size * 8)) { roots + absent <= cells }
+  tot_cells_size:(##(off_bytes * 8))
+  root_list:(roots * ##(size * 8))
+  index:has_idx?(cells * ##(off_bytes * 8))
+  cell_data:(tot_cells_size * [ uint8 ])
+  crc32c:has_crc32c?uint32
+  = BagOfCells;
+```
+Field cells is `n`, `roots` is `k`, `absent` is `l`, and `tot_cells_size` is `Ln` (the total
+size of the serialization of all cells in bytes). If an index is present, parameters
+`s/8` and `t/8` are serialized separately as `size` and `off_bytes`, respectively,
+and the flag `has_idx` is set. The index itself is contained in `index`, present
+only if `has_idx` is set. The field `root_list` contains the (zero-based) indices
+of the root nodes of the bag of cells.
+
+There are also two outdated BoC serialization schemes in the same file.
+
+## SDK from @ton/core
+
+According to the TL-B scheme above there is the [SDK](https://github.com/ton-org/ton-core/blob/main/src/boc/cell/serialization.ts#L1) 
+for serialization and parsing BoC.
+
+Only serialization of BoCs with one root and no absent cells is supported.
+```typescript
+import { beginCell, serializeBoc } from "@ton/core";
+// serializeBoc has two arguments:
+// root: Cell. A root cell of a given tree of cells
+// opt: { idx: boolean, crc32: boolean }. Two flags indicating whether indexes and CRC32C will be included in serialization
+
+const innerCell = beginCell().storeUint(456, 16).endCell();
+
+const rootCell = beginCell().storeUint(0, 64).storeRef(innerCell).endCell();
+
+const serialized_boc = serializeBoc(rootCell, { idx: false, crc32: false });
+
+const serialized_boc_with_indexes_and_crc32 = serializeBoc(rootCell, { idx: true, crc32: true });
```

Comment:
let's make the lengths of the code lines shorter, so everything would fit on the screen without the need to scroll horizontally 

![Uploading telegram-cloud-photo-size-2-5397767000870221510-y.jpg…]()